book/layout.md

```diff
@@ -5,529 +5,632 @@ prev: html
-    self.mr = self.ml = 8
-    self.bt = self.br = self.bb = self.bl = 1
-    self.pt = self.pr = self.pb = self.pl = 8
+def layout(self, tree):
+    # ...
+    self.max_y = document.h
 ```
 
+Then, when the user scrolls down, don't let them scroll past the
+bottom of the page:
+
+``` {.python}
+def scrolldown(self, e):
+    self.scroll = self.scroll + SCROLL_STEP
+    self.scroll = min(self.scroll, self.max_y)
+    self.scroll = max(0, self.scroll))
+    self.render()
+```
+
+Make sure those `max` and `min` calls happen in the right order!
+
 Summary
 =======
 
-In this chapter, we did a pretty dramatic rewrite of the layout
-portion of our browser. We\'ve split layout into two different *layout
-modes* for block and inline content, and we\'ve extended the styling
-capabilities of our browser by adding margins, border, and padding.
+This chapter was a dramatic rewrite of your browser's layout engine.
+That means:
+
+- Layout is now tree-based and produces a *layout tree*
+- Each node in the tree has one of two different *layout modes*
+- Each layout object has a size and position computed
+- The display list now contains generic commands
+- Backgrounds for source code blocks are now drawn.
+
+Tree-based layout makes it possible to dramatically expand our
+browser's styling capabilities. We'll work on that in the [next
+chapter](styles.md).
 
 Exercises
 =========
 
--   Remove the magic numbers from the `Page` object by instead assigning
-    padding to the `<body>` element
--   Add bullets to list items, which in HTML are `<li>` tags. You can
-    make them little squares, located to the left of the list item
-    itself.
--   Add support for background colors and border colors. Code blocks
-    (`<pre>` tags) should have light-gray background color, while
-    headings (`<h2>` tags) should have light-gray bottom borders, or
-    some other colors you like better. Make sure background colors are
-    located behind the text, not in front of it!
--   Implement *margin collapsing*. In margin collapsing, when one block
-    has a bottom margin and the next block has a top margin, the actual
-    gap between them is the *larger* of the two margins, not their sum.
--   Extend margin collapsing to allow the top (bottom) margin of an
-    element to overlap with the top (bottom) margin of its first
-    (last) child, as long as it has no top (bottom) border or padding.
+*Links Bar*: At the top and bottom of each chapter of this book is a
+gray bar naming the chapter and offering back and forward links. It is
+enclosed in a `<nav class="links">` tag. Have your browser give this
+links bar the same light gray background that any other browser does.
+
+*Hidden Head*: There's a good chance your browser is still showing
+scripts, styles, and page titles at the top of every page you visit.
+Make it so that the `<head>` element and its contents are never
+displayed.
+
+*Bullets*: Add bullets to list items, which in HTML are `<li>` tags.
```

Comment:
This one is probably ok, but display:list-item is yet another beast in reality...